Add GHA workflow to perform integration tests

adejanovski opened the pull request:
The GitHub Actions workflow does the following things:
- spins up a 3 nodes cluster in AWS using OSS C* 3.11
- compresses and uploads the diagnostic collection tool to the first node
- runs the collection in extended mode
- downloads the generated tarball locally and decompresses it
- runs the test scenarios written using [Behave](https://behave.readthedocs.io/en/latest/) (Gherkin syntax)
- Tears down the cluster

msmygit commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,94 @@
+
+jobs:
+  integration-tests:
+    runs-on: ubuntu-18.04
```
Do we need to have one for RHEL as well?

adejanovski replied:
it's just the OS which is used to orchestrate the operations, so it doesn't really matter here.
The Cassandra nodes themselves are running on Ubuntu as well, but that's because it's the only OS tlp-cluster can handle (it's not controlled by this setting, but by the AMIs that we use to spin up the instances).
No RHEL/Centos support in tlp-cluster for now.

msmygit commented on tests/integration/__init__.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Spotify AB. All rights reserved.
```
Spotify AB?

adejanovski replied:
:D yep, wrong copyright

adejanovski commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,94 @@
+      if: ${{ success() }}
+      run: |
+          echo "Initializing AWS cluster..."
+          tlp-cluster init DS DSI-58 "Integration tests for the DS diagnostic toolkit" --instance m5.xlarge > /dev/null 2>&1
```
Sure thing, we can definitely spin up smaller instances. I'll take care of that.

adejanovski replied:
m5.large instances are being used now. That's fairly cheap.

adejanovski commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,94 @@
+  integration-tests:
+    runs-on: ubuntu-18.04
+    steps:
+    - uses: actions/checkout@v1
```
sure thing, will do.

adejanovski replied:
Done

adejanovski commented on generate_diag.sh:
```diff
@@ -111,7 +111,7 @@ for i in "$DIAGS_DIR"/${PATTERN}-*.tar.gz; do
 
 if [ -z "$INSIGHTS" ]; then # processing just DSE diagnostic
-    CLUSTER_NAME="$(cat -- */conf/cassandra/cassandra.yaml|grep -e '^cluster_name: '|sed -e "s|^cluster_name:[ ]*\(.*\)\$|\1|"|tr -d "'"|head -n 1|tr ' ' '_')"
+    CLUSTER_NAME=$(cat -- */conf/cassandra/cassandra.yaml|grep -e '^cluster_name: '|sed -e "s|^cluster_name:[ ]*\"\(.*\)\"\$|\1|"|tr -d "'"|head -n 1|tr ' ' '_')
```
The regex which was used kept the `"` that surround the cluster name. This ended up generating tarball with names such as `"Test_cluster"-diagnostics.tar.gz` which turned out to be hard to work with.
The change removes the `"` and generate a proper file name now.
Now that I think of it, some folks may not use double quotes to surround their cluster names as yaml conventions don't enforce it.
I'll rework this to use a simple replacement instead of changing the regex.

adejanovski replied:
I'm hoping to add DSE support to tlp-cluster shortly so that we can create a DSE specific workflow.

adejanovski replied:
regex fixed, I now use `tr` to turn `"` into `_`.

rzvoncek-ds commented on tests/integration/features/steps/integration_steps.py:
```diff
@@ -0,0 +1,74 @@
+# limitations under the License.
+
+
+import configparser
```
A whole lot of these imports is unused.

adejanovski replied:
I cleaned them up, good catch

adejanovski commented:
@rzvoncek-ds @msmygit, I've fixed the PR and added some tests for additional files.

Should be good to review again and hopefully merge.
We should add new tests along the way when we push PRs to the project (coverage is low for now, but the framework is in place at least).

msmygit commented on .github/workflows/oss_unencrypted.yml:
```diff
@@ -0,0 +1,94 @@
+  integration-tests:
+    runs-on: ubuntu-18.04
+    steps:
+    - uses: actions/checkout@v1
```
can we leverage checkout@v2 here?

adejanovski replied:
Done, I had forgotten about it.

msmygit commented on .github/workflows/oss_unencrypted.yml:
```diff
@@ -0,0 +1,94 @@
+          python -m pip install --upgrade pip
+          pip install behave
+          sudo apt-get update
+          sudo apt-get install openjdk-8-jre -y
```
Question: Lately I saw there were some regressions introduced with 275 build of openjdk8. Does that apply here at all? If so, do we need to stick a particular build number?

adejanovski replied:
This u275 issue will affect long running clusters, leaving open sockets piling up.
I don't think that'll affect the tests we're doing here.

If such an issue hit us, we can definitely lock ourselves on a specific JDK version, but for now I guess it's good to test the latest available.
